### PR TITLE
[FIX] Fix executable search

### LINKED ISSUE #98: [BUG] `""` or `''` does not print `PROGRAM_NAME: : command not found`
It does get expanded to an empty string, not `NULL`.
So the issue has to be somewhere after the expansion.

### LINKED ISSUE #110: [BUG] `./<command>` should not convert the exec path to an absolute path
The result should be `./<command>: No such file or directory`.

### LINKED ISSUE #128: [BUG] If cmd_name contains a `/` and is a directory, the error message should be `<cmd_name>: Is a directory`
However, if no `/` is part of the cmd_name, it should print `<cmd_name>: command not found`, even if a directory with this name exists in the current working directory.

### LINKED ISSUE #132: [BUG] `setup_exec_path()` has a bug with `""` (empty string) as input
It will take the first path it finds in the `PATH` environment variable.

### LINKED ISSUE #156: [BUG] Modified `PATH` behavior on execution
No `PATH` environment variable:
```sh
bash-5.1$ ls
bash: ls: No such file or directory
```
Create file called `ls`:
```sh
bash-5.1$ ls
bash: ls: Permission denied
```
Give execute permissions to the `ls` file.
```sh
bash-5.1$ ls
```
Set `PATH` to some garbage:
```sh
bash-5.1$ ls
bash: ls: command not found
```

Whenever there is nothing or just anoher `:` in front or after a `:` in the `PATH` environment variable, it should place a `.` there.
That means `PATH=:`, `PATH=bla1::bla2`, `PATH=:bla` and `PATH=bla:` all result in the following:
```sh
bash-5.1$ ls
bash: ./ls: Permission denied
```

### PR DESCRIPTION
**Pull request by itislu**
#### Building ontop of the knowledge gained in PR #181 and #182 and with help of the source code of bash, this fixes all the exec_path search behaviors, error messages and exit codes.

- If the exit code is **126** (`CMD_EXEC_FAILED`), it means `execve()` failed.
  After that, the error message that gets printed depends on the reason for the failure and depends on `errno`.
- If the exit code is **127** (`CMD_NOT_FOUND`), the process exited before `execve()`.
  This happens if no valid executable file was found. An appropriate error message gets printed before exiting.

---
* Resolves #156 
* Resolves #132
* Resolves #128 
* Resolves #98 
* Resolves #110

### CODE REVIEW

File: include/defines.h [hunk 1]
```diff
@@ -26,7 +26,6 @@
 # include <sysexits.h>
 # include <sys/stat.h>
 # include <sys/wait.h>
-# include <errno.h>
```
Comment (itislu) [inline include/defines.h:29]: Duplicate include

File: .vscode/launch.json [hunk 2]
```diff
@@ -17,6 +17,50 @@
                     "text": "-enable-pretty-printing",
                     "ignoreFailures": true
                 },
+                {
+                    "description": "Follow child processes",
+                    "text": "set follow-fork-mode child",
+                    "ignoreFailures": false
```
Comment (itislu) [inline .vscode/launch.json:23]: This just specifies the default behavior for the debug configuration what process to follow.
It's always possible to go to the other processes if they are waiting on a breakpoint.

File: .vscode/launch.json [hunk 3]
```diff
@@ -17,6 +17,50 @@
...
+                },
+                {
+                    "description": "Detach when child process exits",
+                    "text": "set detach-on-fork off",
+                    "ignoreFailures": false
```
Comment (itislu) [inline .vscode/launch.json:28]: This is what really enables multiprocess debugging.

**Comment by itislu** [comment]
We have to test these test cases of the tester that apparently got worse:
- [ ] 1:
```sh
mkdir tmp_path_test1 tmp_path_test2
printf '#include <unistd.h>\nint main(){write(1, \"1\\n\", 2);}' > tmp_path_test1/test1.c
gcc tmp_path_test1/test1.c -o tmp_path_test1/a.out
printf '#include <unistd.h>\nint main(){write(1, \"2\\n\", 2);}' > tmp_path_test2/test2.c
gcc tmp_path_test2/test2.c -o tmp_path_test2/a.out
export PATH=tmp_path_test1:tmp_path_test2
a.out
export PATH=tmp_path_test2:tmp_path_test1
a.out
/bin/rm -rf tmp_path_test1 tmp_path_test2
```
- [ ] 2:
```sh
ls | wc -l | xargs echo
```

**Comment by itislu** [comment]
It is so weird... No execve fails, and the final_cmd_tables from the main branch and this PR branch look exactly the same.

**Review by LeaYeh — APPROVED** [review]

**Pull request merged**